File: codes/linked_list/reverse_nodes_in_k_group.cpp
#include "../base.h"

 // Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* kNode = getKthNode(head, k);
        if (kNode) {
            ListNode* nextHead = reverseKGroup(kNode->next, k);
            reverseList(head,k);
            head->next = nextHead;
            return kNode;
        }
        else {
            return head;
        }
    }
    
    ListNode* reverseList(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;

        /*
        ListNode* last = head;
        for (int i = 1; i < k; i++) {
            if (!last)
                return head;
            last = last->next;
        }
        */
        
        ListNode* prev = NULL;
        ListNode* cur = head;
        ListNode* t = NULL;
        while (k > 0) {
            t = cur->next;
            cur->next = prev;
 
            prev = cur;
            cur = t;
            
            k--;
        }
        dummy->next = prev;
        
        return dummy->next;
    }
    
    ListNode* getKthNode(ListNode* head, int k) {
        while (k > 1 && head) {
            head = head->next;
            k --;
        }
        
        if (k > 1)
            return NULL;
        return head;
    }
};

File: codes/linked_list/reverse_nodes_in_k_group.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if head is None or k < 2:
            return head
        next_head = head
        for i in range(k-1):
            next_head = next_head.next
            if next_head is None:
                return head
        ret = next_head
        
        current = head
        while next_head:
            tail = current
            prev = None
            for i in range(k):
                if next_head:
                    next_head = next_head.next
                temp = current.next
                current.next = prev
                prev = current
                current = temp
            tail.next = next_head or current
        return ret